pkg/operator/controller/dnsnameresolver-feature/controller.go

```diff
@@ -0,0 +1,160 @@
+		actualName := types.NamespacedName{
+			Namespace: o.GetNamespace(),
+			Name:      o.GetName(),
+		}
+		return expectedName == actualName
+	})
+	if err := c.Watch(source.Kind(mgr.GetCache(), &configv1.FeatureGate{}), &handler.EnqueueRequestForObject{}, clusterNamePredicate); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// Config holds all the configuration that must be provided when creating the
+// controller.
+type Config struct {
+	// DNSNameResolverEnabled indicates that the "DNSNameResolver" featuregate is enabled.
+	DNSNameResolverEnabled bool
+
+	FeatureGateName string
+
+	// DependentCaches is a list of caches that are used by Controllers watching DNSNameResolver
+	// resources. The dnsnameresolverfeature controller starts these caches once
+	// the DNSNameResolver CRD has been created.
+	DependentCaches []cache.Cache
+	// DependentControllers is a list of controllers that watch DNSNameResolver
+	// resources. The dnsnameresolverfeature controller starts these controllers once
+	// the DNSNameResolver CRD has been created and the DependentCaches are started.
+	DependentControllers []controller.Controller
+}
+
+// reconciler handles the actual featuregate reconciliation logic in response to
+// events.
+type reconciler struct {
+	config Config
+
+	cache            cache.Cache
+	client           client.Client
+	startCaches      sync.Once
+	startControllers sync.Once
+}
+
+// Reconcile expects request to refer to a FeatureGate and creates or
+// reconciles the DNSNameResolver CRD.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	logrus.Info("reconciling", "request", request)
+
+	// Check if the DNSNameResolved feature gate is enabled or not.
+	if !r.config.DNSNameResolverEnabled {
+		return reconcile.Result{}, nil
+	}
+
+	// Check if the name of feature gate matches with the default feature gate.
+	if request.Name != r.config.FeatureGateName {
+		// Return a nil error value to avoid re-triggering the event.
+		logrus.Errorf("skipping unexpected feature gate object %s", request.Name)
+		return reconcile.Result{}, nil
+	}
+
+	// Get the feature gate object.
+	featureGate := configv1.FeatureGate{}
+	if err := r.cache.Get(ctx, request.NamespacedName, &featureGate); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Check that the feature set is either TechPreviewNoUpgrade or CustomNoUpgrade.
+	if featureGate.Spec.FeatureSet != configv1.TechPreviewNoUpgrade && featureGate.Spec.FeatureSet != configv1.CustomNoUpgrade {
+		return reconcile.Result{}, nil
+	}
+
+	// Ensure the DNSNameResolver CRD is created for the enabled feature set.
+	if err := r.ensureDNSNameResolverCRD(ctx, featureGate.Spec.FeatureSet); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Start the dependent caches and wait for the caches to sync.
+	r.startCaches.Do(func() {
```

Comment:
I don't think `startCaches` or `startControllers` should be called in the reconcile.  Shouldn't it be called (and watched) in `New`?

Reply:
The caches and controllers are started only if the DNSNameResolver feature is enabled. The controllers are unmanaged, so these are not started when the operator starts. Once the dnsnameresolvercrd controller starts, it starts the dependent controllers if the feature is enabled. The dnsnameresolver controller uses different caches for storing the objects it is watching. These caches are needed to be started before the controller is started. 

Similar approach is used in cluster-ingress-operator for gateway api: https://github.com/openshift/cluster-ingress-operator/blob/9a872878deb0421105da88c9fe25b1079aa2f2ea/pkg/operator/controller/gatewayapi/controller.go#L91-L100

Reply:
The gatewayapi controller's  purpose is to watch the feature gate and update the Gateway API CRDs as a part of the installation process for Istio Gateway API.  This controller does more. Maybe splitting the cache handling into the other dnsresolver controller is the solution.

Reply:
Also, aren't your CRDs in the OpenShift API?  The Gateway API CRDs are not a part of the OpenShift API so this is a little different situation.

Reply:
> The gatewayapi controller's purpose is to watch the feature gate and update the Gateway API CRDs as a part of the installation process for Istio Gateway API. This controller does more. Maybe splitting the cache handling into the other dnsresolver controller is the solution.

This controller (`dnsnamersolverfeature` controller) also does the same thing. It watches the feature gate and creates/updates the DNSNameResolver CRD. Along with that it starts the `dnsnameresolver` controller and the caches which are used by that controller. The caches are needed to be started before the `dnsnameresolver` controller or else the controller will not be able to function properly (it might throw an error as well). The difference with Gateway API controller is that the [dependent controllers](https://github.com/openshift/cluster-ingress-operator/blob/9a872878deb0421105da88c9fe25b1079aa2f2ea/pkg/operator/operator.go#L302-L311) that it starts uses the same cache ([gatewayclass controller](https://github.com/openshift/cluster-ingress-operator/blob/1938e2cbf206769e6195729ee751443b5c0afdaa/pkg/operator/controller/gatewayclass/controller.go#L47) & [gateway service dns controller](https://github.com/openshift/cluster-ingress-operator/blob/1938e2cbf206769e6195729ee751443b5c0afdaa/pkg/operator/controller/gateway-service-dns/controller.go#L62)) that the operator manager creates wheres the caches used by `dnsnameresolver` controller are different as one cache watches the DNSNameResolver objects in `openshift-ovn-kubernetes` namespace and the other one watches CoreDNS pod endpointslices in `openshift-dns` namespace.



> Also, aren't your CRDs in the OpenShift API? The Gateway API CRDs are not a part of the OpenShift API so this is a little different situation.

I am not sure about the difference apart from the CRDs reside in different repos. Can you please explain?